### ISSUE TITLE
rapid dev for local unit testing

### ISSUE THREAD

**Issue by mattdesl** [hughsk/smokestack#45]
So I'm basically craving [nodemon](https://www.npmjs.com/package/nodemon) but running in a browser, so I can do rapid TDD on web APIs like WebGL, WebAudio, etc. If something like this already exists, let me know.

I'm pretty close to a nicely working solution using [budo](https://www.npmjs.com/package/budo) (though any similar tool could work) and tap-closer:

```js
var smokestack = require('smokestack')
var file = require.resolve('tap-closer/bundle.js')
var closer = require('fs').readFileSync(file)

var entries = ['tests/test-texture-2d']
require('budo')(entries)
  .on('update', function(file, contents) {
    var stream = smokestack()
    stream.pipe(process.stdout)
    stream.write(closer)
    stream.end(contents)
  })
```

:tada: 

On filesave, this re-runs the tests and pipes them through smokestack into terminal.

![screen shot 2015-05-27 at 10 32 43 pm](https://cloud.githubusercontent.com/assets/1383811/7851584/603cb73c-04c0-11e5-9148-3c2c2f4b36be.png)

Basically my questions come down to:

- Is there a way to not show the smokestack browser window, or make it much smaller, or less intrusive?
- Is there a better way of achieving this than using smokestack?

**Comment by mattdesl** [comment]
Oh! I was just reading another issue #29 and I probably don't need budo at all, and can just stick to watchify directly for this.

**Comment by hughsk** [comment]
Nothing in place at the moment, but there should definitely be a way to hide the window somewhere amidst Chrome's CLI and clientside JS – it'd be awesome to have a `--hide` flag as the browser popup is pretty distracting.

That script you posted looks pretty nice and clean for a bunch of projects. Long-term the goal was to split out the browser instrumentation stuff into separate modules so that we could build more complex tools without complicating smokestack too much. Sadly none of us have had the chance to get round to it yet... Though of course we're open to PRs :)

Otherwise [zuul](https://github.com/defunctzombie/zuul) could be worth investigating in the meantime? It's a little longer to set up but may have stuff in place for those situations where you want rapid iteration like that.

**Comment by mattdesl** [comment]
I've taken a look at zuul before, don't know if it's exactly what I want. But I should probably give it another shot.

Since posting this I've been tinkering with budo, chrome-launch and chrome-remote-interface to reload the page and capture logs. It reloads insanely fast and the window can be used as another test visualizer (see tap-browser-colors module). 

Regular tap works great. Some things I still need to figure out:

- tap-spec isn't receiving an "end" event so it never totals the failures/successes 
- on errors, tap-spec is trying to readFileSync the bundle.js (there is no local file)
- no source maps on errors :(

**Comment by hughsk** [comment]
Oooh nice, look forward to seeing the results. I think @timoxley has delved deep into getting sourcemaps working properly in the past and might have some insight there. There's [source-map-support](https://github.com/evanw/node-source-map-support) which has worked for me sometimes but not always.

Hmm, the tap-spec issues might call for something to use in its place?

**Comment by mattdesl** [comment]
So.. there is Chromium switch `--no-startup-window` but unfortunately it won't actually run any JS, so it's pretty useless. There is also `--window-size=w,h` which helps make things a little less intrusive. In an ideal world the tool would just open DevTools Console window instead of a blank browser window.

I'll keep poking around, and probably use `shoe` like smokestack to relay console logs.

**Comment by paulirish** [comment]
If you want devtools to open then you could set up remote-debugging-port and sniff out the URL for your inspection. That's not perfect, but it's close.

Here's a related ticket around a flag to force devtools open https://code.google.com/p/chromium/issues/detail?id=410958  It's not implemented yet.

**Comment by mattdesl** [comment]
Cheers @paulirish. I'm launching Chrome with remote debugging, opening two tabs: `localhost:9966` and `localhost:9222/json`. But I only see the `devtoolsFrontendUrl` in the JSON once I manually open dev tools on the `localhost:9966` tab.

**Comment by mattdesl** [comment]
I've been looking into Electron for this; it's a bit heavy but it works pretty well for my use case.

https://github.com/Jam3/hihat

:tada:

**Comment by yoshuawuyts** [comment]
Neat!

**Comment by paulirish** [comment]
hihat is so rad.

**Comment by mattdesl** [comment]
thanks :beers:
